File: typescript/src/problems1_100/palindrome-number.ts
class Problem {
  isPalindrome(x: number): boolean {
    throw new Error(`Not implemented (${x}).`);
  }
}

class ReverseString extends Problem {
  isPalindrome(x: number): boolean {
    const s: string = x.toString();
    return s === s.split('').reverse().join('');
  }
}

class CompareHalfString extends Problem {
  isPalindrome(x: number): boolean {
    const s: string = x.toString();
    let start: number = 0;
    let end: number = s.length - 1;
    while (start < end) {
      if (s[start++] !== s[end--]) {
        return false;
      }
    }
    return true;
  }
}

const SOLUTIONS = [new ReverseString(), new CompareHalfString()];
export default SOLUTIONS;

File: python/src/problems1_100/palindrome_number.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def is_palindrome(self, x: int) -> bool:
        pass


class ReverseString(Problem):
    @override
    def is_palindrome(self, x: int) -> bool:
        s: str = str(x)
        return s == s[::-1]


class CompareHalfString(Problem):
    @override
    def is_palindrome(self, x: int) -> bool:
        s: str = str(x)
        start: int = 0
        end: int = len(s) - 1
        while start < end:
            if s[start] != s[end]:
                return False
            start += 1
            end -= 1
        return True


SOLUTIONS: list[Problem] = [ReverseString(), CompareHalfString()]
